packages/store/addon/-private/system/store/record-data-store-wrapper.ts

```diff
@@ -188,17 +208,9 @@ export default class RecordDataStoreWrapper implements IRecordDataStoreWrapper {
       throw new Error(MISSING_ID_ARG_ERROR_MESSAGE);
     }
 
-    let internalModel = internalModelFactoryFor(this._store).peekId(modelName, id, clientId);
+    let internalModel = internalModelFactoryFor(this._store).peek(modelName, id, clientId);
     if (internalModel) {
       internalModel.destroyFromRecordData();
     }
   }
 }
-
-function hasValidId(id?: string | null, clientId?: string | null): id is string {
```

Comment:
This is a questionable type guard, and if misused could result in some "lying" in the type-checker. There's a significant difference between the static analysis behavior

`"true if id is a string"`

and the runtime behavior

`"true unless both id and clientId are truthy"`

We should take a close look at this before merging the PR

Reply:
We've looked at this together before. It's basically a "caught between a rock and a hard place" situation where method signature overloads don't "fall through" to calls to other methods with similar signatures.  Here's an example https://typescript-play.js.org/#code/GYVwdgxgLglg9mABAEzgdQBQEMAqALGMAcwC5EBnKAJ0KIBpEsw4o8BTK-Wsym4xAD6IwIADaiAlGQBucGMgDcAKFCRYCFOmxdiParUHCxohkxbtOBXRX3EpiWfOWro8JKky4rpG3yKGRcVNmVg4dH14DIUDJGTlkRABvAF8lFXBXDVQADW1vPT9g8zD83yijcXtHRXS1N01cr24y-mjjItDLZsi7OKdazPc4RvCC8piOi1GW-zbKvoTEpUQVxAB6NcQOKjgqckQAIzYILBByNkZwxBh9gAMe2YrRW+XVqjYoECohz3DJktoEmUqQG6iGAE08t1bPRGCEpqUHgFjFV4s4MmDNJCmtYkXMTHDil1cTDUf0XJjUNjpninv9iREYcj5g54klXqsNogAEQAd12AGtyNyOYh3p9vppft5GPsHvTwkClCCKfVUAAtKEkwqEzo0pn4sk1VVZOCanGMvzMglmPWI0kLdF1U3m-VW-EK+3up5G9mrdabPmC4Wi8VfH5a2G2hEGLBy0nAtJcoNUIXcw4gKDCOC8xiIUQsRBwYCICBULAALwAnqX2BAhYx3tnc+8AI4gGDvBKERBQBD7YuIAC2HzwcGQ5CUXN57CQ3IQF0HoXOve8+yOsZm6d5cYcWFE8iwUDYCSw+YQ-lgI8ObGAuzYoLVcAAwuXq8+60LI2NiJ7od6YiNJ1Bk0V9KyrD9ji-C0f0eCZdRjbVAUdR9TTA99P3Ib8ZmtP8kNaH0Fj9VYYBLbB4QBfgAEIAF4aJ9Yj-TDSUPEjPCiCVf1ki2UQV1IxBI0QWj6MAxi3g+cMpTYhDKI45QuJ4lcln9FYuTANhpA4UUVlYHZc3U3MAFEqB2KgMFuAA5DSOFuTjVlSf1lSAA

Reply:
I would also note that the near-term solution for not needing this craziness is to continue refactoring towards use of `identifiers`. Because identifiers are typed as a union of these two cases (one or the other must be a string) we reduce the number of method overloads and pass the "overloaded" type signature around with the value instead.